runtime/pkg/rilltime/rilltime_test.go

```diff
@@ -1,176 +1,576 @@
+	grains    = []string{"s", "m", "h", "D", "W", "M", "Q", "Y"}
+	now       = "2025-05-15T10:32:36Z"
+	minTime   = "2020-01-01T00:32:36Z"
+	maxTime   = "2025-05-14T06:32:36Z"
+	watermark = "2025-05-13T06:32:36Z"
+
+	twoPeriodAgoStarts = []string{
+		"2025-05-13T06:32:34Z", // s
+		"2025-05-13T06:30:00Z", // m
+		"2025-05-13T04:00:00Z", // h
+		"2025-05-11T00:00:00Z", // D
+		"2025-04-28T00:00:00Z", // W
+		"2025-03-01T00:00:00Z", // M
+		"2024-10-01T00:00:00Z", // Q
+		"2023-01-01T00:00:00Z", // Y
+	}
+	twoPeriodAgoWeekBoundaryStarts = []string{
+		"2025-05-13T06:32:34Z", // s
+		"2025-05-13T06:30:00Z", // m
+		"2025-05-13T04:00:00Z", // h
+		"2025-05-11T00:00:00Z", // D
+		"2025-04-28T00:00:00Z", // W
+		"2025-03-03T00:00:00Z", // M
+		"2024-09-30T00:00:00Z", // Q
+		"2023-01-02T00:00:00Z", // Y
+	}
+	prevPeriodStarts = []string{
+		"2025-05-13T06:32:35Z", // s
+		"2025-05-13T06:31:00Z", // m
+		"2025-05-13T05:00:00Z", // h
+		"2025-05-12T00:00:00Z", // D
+		"2025-05-05T00:00:00Z", // W
+		"2025-04-01T00:00:00Z", // M
+		"2025-01-01T00:00:00Z", // Q
+		"2024-01-01T00:00:00Z", // Y
+	}
+	curPeriodStarts = []string{
+		"2025-05-13T06:32:36Z", // s
+		"2025-05-13T06:32:00Z", // m
+		"2025-05-13T06:00:00Z", // h
+		"2025-05-13T00:00:00Z", // D
+		"2025-05-12T00:00:00Z", // W
+		"2025-05-01T00:00:00Z", // M
+		"2025-04-01T00:00:00Z", // Q
+		"2025-01-01T00:00:00Z", // Y
 	}
+	curPeriodEnds = []string{
+		"2025-05-13T06:32:37Z", // s
+		"2025-05-13T06:33:00Z", // m
+		"2025-05-13T07:00:00Z", // h
+		"2025-05-14T00:00:00Z", // D
+		"2025-05-19T00:00:00Z", // W
+		"2025-06-01T00:00:00Z", // M
+		"2025-07-01T00:00:00Z", // Q
+		"2026-01-01T00:00:00Z", // Y
+	}
+)
 
-	for _, testCase := range testCases {
-		t.Run(testCase.timeRange, func(t *testing.T) {
-			rt, err := Parse(testCase.timeRange, ParseOptions{})
-			require.NoError(t, err)
+func Test_CompletePreviousAndCurrentGrain(t *testing.T) {
+	var testCases []testCase
 
-			start, end, grain := rt.Eval(EvalOptions{
-				Now:        now,
-				MinTime:    minTime,
-				MaxTime:    maxTime,
-				Watermark:  watermark,
-				FirstDay:   1,
-				FirstMonth: 1,
-			})
-			require.Equal(t, parseTestTime(t, testCase.start), start)
-			require.Equal(t, parseTestTime(t, testCase.end), end)
-			require.Equal(t, testCase.grain, grain)
+	expectedGrains := []timeutil.TimeGrain{
+		timeutil.TimeGrainMillisecond, // for s
+		timeutil.TimeGrainSecond,      // for m
+		timeutil.TimeGrainMinute,      // for h
+		timeutil.TimeGrainHour,        // for D
+		timeutil.TimeGrainDay,         // for W
+		timeutil.TimeGrainDay,         // for M
+		timeutil.TimeGrainMonth,       // for Q
+		timeutil.TimeGrainMonth,       // for Y
+	}
+
+	for i, grain := range grains {
+		tmGrain := timeutil.TimeGrain(i + 2)
+
+		// Current period
+		testCases = append(testCases, testCase{
+			timeRange: fmt.Sprintf("%[1]s^ to %[1]s$", grain),
+			start:     curPeriodStarts[i],
+			end:       curPeriodEnds[i],
+			grain:     expectedGrains[i],
+		})
```

Comment:
The automation here has made these test cases hard to follow. They are also no longer useful as examples we can easily reference when writing docs or doing support.

Even if it means some repetition, I favor the previous test syntax that clearly shows how an expression evaluates:
```
{"W1 to W3", "2025-03-03T00:00:00Z", "2025-03-17T00:00:00Z", timeutil.TimeGrainDay},
{"W1 to W3 by W", "2025-03-03T00:00:00Z", "2025-03-17T00:00:00Z", timeutil.TimeGrainWeek},
...
```

Reply:
Sounds good. Lemme add explicit entries.